### ISSUE TITLE
MathJax disappearing overline and left/right abs bug effects some users on my WW 2.16 server

### ISSUE THREAD

**Issue by taniwallach** [openwebwork/webwork2#1524]
MathJax 3.1.2 (used in WW 2.16) seems to have an issue with `\overline{}` disappearing from the screen for some browsers (apparently Chrome and Firefox) in certain circumstances.

---

Edit  (04-Apr-2022): Issues of disappearing `\left\abs` and `\right\abs` have also been encountered in both Chrome and Firefox.

Edit (04-Apr-2022): Using the SVG output mode of instead of `chtml` MathJax seems to avoid these issues.

---

For example some of students using my system have reported that a `\overline{}` was not being shown in problems with the following equations:
```
\( \displaystyle \frac{ z_{1}^{3} - (\overline{z_1})^3 \, }{z_2} = \)
```
and
```
\( \left| \, \left( z_1 \cdot z_2 - \overline{z_3} \, \right)^3 \, \right| = \)
```

The issue seems to be specific to certain browsers and zoom levels.

Sometimes the line appears/disappears after a page refresh or problem submission according to the reports. (I was sent a video of the disappearing `\overline` after a submit.)

I don't seem to be able to trigger the issue on my machine (Debian 11 with Chrome version 95.0.4638.69) except at 50% zoom on Chrome which is so small it cannot be read easily anyway. 

Firefox 78.15.0esr (Debian's packaged version) does not seem to hit it on my machine for any standard zoom level. 

---

Edit (04-Apr-2022): More recent versions of Firefox (94-98) certainly do encounter these issues.

---

The problem seems to be a "quirk" with clipping behavior, and @dvpc seems to be trying to have MathJax work around the issue. It is addressed in:
  - https://github.com/mathjax/MathJax/issues/2701   (refers to chtml output, as WeBWorK uses)
  - https://github.com/mathjax/MathJax-src/pull/726
The reports indicate that `\underline{}`is also effected.

There are quite a few OPL questions which use `\overline{}` so trying to avoid the issue probably justifies a hotfix for WW 2.16 once a reasonable solution is found.  https://github.com/openwebwork/webwork-open-problem-library/search?q=overline

Upgrading to MathJax 3.2.0 may fix the problem, but that is not certain and may be more of a change than we want in a hotifx. 

https://github.com/mathjax/MathJax/issues/2795
reports the problem still occurring for Firefox for some zoom levels and OS versions.

A temporary fix by using the CSS workaround by @dpvc in https://github.com/mathjax/MathJax/issues/2701#issuecomment-859594360 might be a simpler fix for WW 2.16, if it works well enough.

@Alex-Jordan posted https://github.com/mathjax/MathJax/issues/2701#issuecomment-858121404 that he could trigger the issue for certain zoom levels on Chrome 91 on MacOS with the sample in the MathJax issue.

@pstaabp @mgage @drgrice1 - Do any of you encounter the bug on your machines with WeBWorK 2.16 servers and formula like the ones above?

**Comment by Alex-Jordan** [comment]
If you would like to upgrade to MathJax 3.2.0, try the steps here:

https://webwork.maa.org/moodle/mod/forum/discuss.php?d=6374https://webwork.maa.org/moodle/mod/forum/discuss.php?d=6374

In my case, I first manually edited `htdocs/package.json` and `htdocs/package-lock.json` to have `3.2.0` instead of `3.1.2`. Maybe that is needed.

**Comment by Alex-Jordan** [comment]
And beware of caching. If you think you have moved to MJ 3.2.0, use the MJ menu to view the version to double check.

**Comment by taniwallach** [comment]
It does not seem that upgrading to MathJax 3.2.0 will fix the disappearing overline problem for recent versions of Firefox. (I tested only with regards to the CSS patch and not with a full upgrade of MathJax.)

@maharper - Can you test the sample problem from below using only the `overflow: clip visible ! important;` CSS addition on whatever machine you tested with before. I assume the results will be similar to what I saw on Linux. I don't think there is much point in testing the ugly "scaling" hack at all - it breaks `\overrightarrow` so really only serves to demonstrate that the font-size has an effects on when the bug occurs in Firefox.

It certainly seems that the root cause may be a bug in recent versions of Firefox, in particular given that  Firefox 78esr seems to handle the overline better than Firefox 94.

It could be that some other patch to the MathJax CSS may help - but I'm not sure what else to try.

Right now - the situation is that I am not certain what would fix the problem for recent versions of Firefox.

---

With a more recent version of Firefox (94.0.1) installed via FlatPak on my machine (Debian 11) - I can see the issue with the default MathJax of WeBWorK 2.16.

Testing in Firefox was done with a problem file based on the one at the bottom of this post.

The behavior in Firefox 94.0.1 is quite unusual.
  - Different "extenders" appeared/disappeared at different zoom levels.
  - The results with vs. without `overflow: clip visible ! important;` are such that the extra CSS sometimes helped (ex. at 110%) and sometimes made things worse (ex. at 100%).
  - **With no extra CSS added** (the entire `TEXT()` block which added the CSS code was removed).
    - At 100% just the "overline portion" of an `\overrightarrow` was missing (far right term) but everything else appeared.
    - At 110% the "overline" were missing except the "overline portion" of an `\overrightarrow` which did appear as did the `\underline`. 
    - At 120% only the `\underline` was missing.
    - At 133% everything appeared.
    - At 150% everything appeared.
    - At 90% different results on different returns to this zoom level (or there was a mistake in my notes):
      - Sometime just the "overline portion" of an `\overrightarrow` was missing (far right term) but everything else appeared.
      - Sometime just the "overline portion" of an `\overrightarrow` appeared but everything else was missing.
    - At 80% all the overlines including except that of the `\overrightarrow` were visible; the `\underline` was visible. 
    - At 67% just the "overline portion" of an `\overrightarrow` appeared.
    - At 50% all the extenders were missing.
  - **With just the `overflow: clip visible ! important;` added via local CSS**:
    - At 100% only the "overline portion" of an `\overrightarrow` appeared  (far right term); all the other "extenders" were missing
    - At 110% everything appeared.
    - At 120% everything appeared.
    - At 133% and 150% the `\overlines` were missing, but the full `\overrightarrow` appeared as did the `\underline`
    - At 170% everything appeared.
    - At 90% everything seems to appear.
    - At 80% all the extenders were missing.
    - At 67% the "overline portion" of the `\overrightarrow` appeared, as did the `\underline`
    - At 50% all the extenders were missing.

My conclusion:
  - For Firefox 94 (on Linux) the patch `overflow: clip visible ! important;` does not consistently improve the output.

With the old Firefox 78.15.0esr from Debian, the only missing extender I noticed in these problems was the `\underline` at 67% which was missing both with and without the CSS patch.

When the same 2 options were tested on Chrome (Version 95.0.4638.69 on Debian 11):
  - From 67% zoom and up all the extenders seem to appear. At 50% some of the lines seem grey instead of black.
  - At 50% most extenders are missing without the additional CSS, and with the additional CSS it all appears.
  - At 33% on my monitor the extenders are mostly missing, except the part for the `\overrightarrow` but overall the math is essentially too small to read.

I also tried testing quickly with the suggestion from https://github.com/mathjax/MathJax/issues/2701#issuecomment-857641697 but that did not seem to help.

Below is a sample problem file which uses local CSS code to try to overcome the disappearing `\overline` problem, and allows tweaking the settings. It includes some values which seem to get `\overline` to behave more-or-less reasonably. However, that change breaks `\overrightarrow`. The "hack" uses a larger scale of the font and shifts the "box" down. I fiddled with the values of `font-scale` and `top` to get something which worked for me at several scale value in FireFox. However, this approach breaks `\overrightarrow` as the arrow-head does not "attach" to the `\overline` based "tail" of the arrow.

The red coloring set in the CSS was just to draw attention to the overline for the testing.


```
DOCUMENT();      

loadMacros( "PGstandard.pl" );

TEXT( MODES(
  HTML=>"
    <style>
    mjx-over > mjx-mo > mjx-stretchy-h > mjx-ext {
      color: red ! important;
      font-size: 240%;
      overflow: clip visible ! important;
      position: relative ! important;
      top: 0.3em ! important;
    }
    </style>",
    TeX=>""
));


BEGIN_TEXT
\( \displaystyle \frac{ z_{1}^{3} - (\overline{z_1})^3 \, }{z_2}
\quad
\left| \, \left( z_1 \cdot z_2 - \overline{z_3} \, \right)^3 \, \right| 
\quad
P\left( \overline{A \cup B \cup C} \right)
\quad
P\left( \underline{A \cup B \cup C} \right)
\quad
P\left( \overrightarrow{A \cup B \cup C} \right)
\)
$BR

END_TEXT

ENDDOCUMENT();        

```

**Comment by taniwallach** [comment]
Bug report filed with Mozilla:   https://bugzilla.mozilla.org/show_bug.cgi?id=1741887

**Comment by maharper** [comment]
@taniwallach I observe much the same using 91.3.0esr on Windows 10.  The issue of the disappearing overlines is not resolved by Mathjax 3.2.0 nor by the CSS
```[css]
mjx-over > mjx-mo > mjx-stretchy-h > mjx-ext {
      overflow: clip visible ! important;
    }
```

**Comment by taniwallach** [comment]
I still am not  aware of any reasonable fix for this issue which helps with recent versions of FireFox. It seems that it is probably a browser issue and not something which can be easily fixed on the MathJax or WeBWorK side.

In the few questions where `\overline` is currently in use on my server - we put a note about the issue explaining where it should be. That is not a reasonable long-term solution, nor feasible if there are many affected questions.

**Comment by dpvc** [comment]
It turns out that this is not a clipping problem, but a Firefox rendering issue for think horizontal lines in connection with the `transform` CSS that is used to make the extenders.  At some resolutions, these thin lines are dropped (probably due to round-off issues when trying to match the pixel positions on screen).  The following styles worked for me in Ubuntu 20.04 with Firefox 95.0:

``` css
mjx-stretchy-h.mjx-cAF mjx-ext mjx-c::before {
  transform: scale(500,1.9);
  transform-origin: center top;
}
mjx-stretchy-h.mjx-cAF mjx-ext {
  overflow: clip visible
}
```

This is specific to MathJax v.3.1.2, so would need to be changed for MathJax v3.2, but since 3.1.2 is what is used by WW2.16, that is what I used for testing.

See if that works for you.  I wasn't able to reproduce the problem in Windows 10 (I have a pretty old version and didn't take time to get a new one) with Firefox 95.0.1.  I also wasn't able to reproduce it with Firefox 95.0.1 in Mac OS X, even in the situations that used to be problematic in MathJax v3.1.2, so I'm wondering if Firefox 95 handles this better (though still not perfectly in Ubuntu at least).

**Comment by taniwallach** [comment]
I tested with Firefox 95.0.2 (flatpack) on Debian 11. My tests are not conclusive.

There seems to be some improvement for some zoom levels and things seem worse at others. 

  - 100% scale - both with and without the extra CSS, the left straight part of `\overrightarrow` is missing, as in the `\underline`
  - 90% scale - with the extra CSS some lines seem "faded"
  - 80% scale - many lines are missing both with and without the extra CSS
  - 67% scale - things are better without the extra CSS
It seems that trying to get the Mozilla people to look into the issue might be the best course of action, but they do not seem to think this is of any significant priority.

**Comment by dpvc** [comment]
The CSS in my message above should ONLY affect the `\overline`, and should affect `\overrightarrow`.  If `\overrightarrow` is also problematic, then that would require additional CSS.

**Comment by taniwallach** [comment]
> Bug report filed with Mozilla: https://bugzilla.mozilla.org/show_bug.cgi?id=1741887

People related to the Firefox project have looked into the bug report, and have acknowledged that it is a bug. It was determined that it can affect rendering of other very thin lines.

Unfortunately the bug was assigned what is apparently a low priority and severity.

People who find the bug makes trouble for them or their students are invited to vote for the bug.

**Comment by taniwallach** [comment]
There are several reports of similar problems with thin vertical lines in recent versions of Firefox.

I now have a report of a disappearing `\left\abs` in Chrome.

At present, the best advice I have for affected users is to either use a different browser or to set "images" mode for the equations (which is far from ideal). It seems that the SVG output mode avoids these issues, but at present there is no option I am aware of to allow a user/course to request using the SVG output mode of MathJax instead of CommonHTML (HTML+CSS).

**Comment by taniwallach** [comment]
Bottom line: The solution (at present) to this problem is to set WeBWorK to use MathJax's SVG output format. That has been done by default for the newly released WeBWorK 2.17 in a hot-fix.

In the future, either a Firefox bug fix or changes to the MathJax fonts may solve this display problem in Firefox.
